### PR TITLE
Feature cli polish

### LINKED ISSUE #44: CLI polish and fixes for deployment
### Specification

While testing hole punching with the testnet, we found a few issues with the CLI.

1. Nodes Get All is throwing an error
2. We are getting a weird output format for nodes connections what is the 0 and what is this NA for?
3. I think nodes connections is better suited for dictionary format - list format or table format (can you check which ones done robustly)
4. The output changes sometimes, to just join the NA after the ip address, so no port is available
5. Summary - the nodes connections command is just very non-deterministic
6. Secret leak is occurring on agent service, there should be no authorisation metadata see --format=json
8. What's usageCount

### Additional context
 
* Related: https://github.com/MatrixAI/Polykey/issues/551

### Tasks

- [x] 1. Fix up `nodes getall` command, it's throwing an error relating to `host` not being defined.
- [x] 2. Review and fix up the `nodes connections` command output formatting. The output is very unclear.
- [x] 3. client call metadata is being included in the `nodes connections --format json` output. This should not be the case.
- [x] 4. `agent status` command outputs too much information. We should strip out the key and certificate information.

### PR DESCRIPTION
**Pull request by addievo**
### Description

This PR aims to fix a lot of UI/UX issues of Polykey CLI to improve clarity and security of command outputs.

The main issues to be fixed include: 

1. Fixes `nodes getall`:

This previous threw an error related to undefined host, this has now been resolved in the corresonding Polykey Handler by getting the valid address and port from the `info` field which is of type `NodeData`
```ts
          host: info.address.host,
          port: info.address.port,
```

2. Improve `nodes connections` Output: 

Previously, the output was rather unclear and lacked corresponding labels to relevant labels for `Node Host`, `UsageCount` and `Timeout`, now labels are added for the same.

Previously we used a hardcoded amount of tabbed spaces between columns, this has now been changed to be padded based on length of row with the max length.

```ts
// Old Code
const outputLine = `${hostString}\t${usageCount}\t${timeout}`;

// New Code
const outputLine = hostString.padEnd(maxHostStringLength) + "\t" + usageCount.padEnd(maxUsageCountLength) + "\t" + timeout.padEnd(maxTimeoutLength);
```

3. Sanitizing `JSON` formatted output for `nodes connections`:

This was a critical security issue where in the sensitive information of authentication data (although this is the user's authentication data itself) was being included in the JSON output, this has been sanitized out.

```ts
// Implementation Snippet
const sanitizedConnections = JSON.parse(JSON.stringify(connections));
delete connection.metadata;
```

4. Refactor `agent status` output:

JWK and PEM have very little usage in status, and correspondingly are being removed from agent status, this is a very simple fix of just removing the following keys from the output

```ts
// Old Code
data: { ... , publicKeyJWK: response.publicKeyJwk, certChainPEM: response.certChainPEM }

// New Code
data: { ... , /* publicKeyJWK and certChainPEM removed */ }
```

5. Additional scope added to this PR - Standardisation of table layout for `outputFormatter` : 

Table output in `outputFormatter` was intiially very static and did not take into account variety in terms of padding, this is altered now to allow field for headers, and a boolean to toggle numbering of rows. 

`Table` format now includes a standardised TSV padding.

`Table` format handles `\t` and `\n` by enclosing them in quotation marks to prevent disruption of the layout.

Additional supported is added for stream based padding to improve efficiency to `O(2n)`.

### Issues Fixed
* Related #44 

### Tasks
- [x] 1. `nodes connection` human format label.
- [x] 2. `nodes connection` json format remove metadata (sanitize).
- [x] 3. Remove `JWK` and `PEM` from `agent status`.
- [x] 4. Add a jest for `nodes connection`.
- [x] 5. Update jests which rely on `agent status`.

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

### CODE REVIEW

**Comment by addievo** [comment]
Sanitization isn't a good solution for num 2. 

This issue is related to this issue:

https://github.com/MatrixAI/Polykey/issues/572

**Comment by addievo** [comment]
Console output from handler for number 2:

```ts
{
  nodeId: IdInternal(32) [Uint8Array] [
    211, 206, 216, 221, 137,  66, 215, 114,
      8, 228,  98, 250, 193, 250, 150,  56,
    210,  34,  94, 186,  71, 180, 100, 130,
    229, 191, 254,  19,  45,  18, 227, 171
  ],
  address: { host: '127.0.0.1', port: 55551, hostname: undefined },
  usageCount: 0,
  timeout: undefined
}
```

So it is more than obvious that this JSON:

```JSON
[{"host":"127.0.0.1","hostname":"","nodeIdEncoded":"vqf7dhnc98bbn4274cbtc3ukm73924nlq8uq690n5nvv16b8iselg","port":55551,"timeout":-1,"usageCount":0,"metadata":{"timeout":15000,"authorization":"Bearer {\"payload\":\"eyJpYXQiOjE2OTgyMTY4NTguNzA1LCJpc3MiOiJ2cTQ4MG9mbGQ3dnJvbDB1cmExdmlpZjk0NzRwYjdwdHJxbTlzOGxkMjN2bjV0OTJvaWY0ZyIsInN1YiI6InZxNDgwb2ZsZDd2cm9sMHVyYTF2aWlmOTQ3NHBiN3B0cnFtOXM4bGQyM3ZuNXQ5Mm9pZjRnIn0\",\"signatures\":[{\"protected\":\"eyJhbGciOiJCTEFLRTJiIn0\",\"signature\":\"UbIRKl5a8AWuNIAK2aCRmgatv0e940ZNZJ2Zvzi_WF4\"}]}"}}]
```

The metadata field isn't from the backend, consequently the issue lies in the frontend.

**Comment by addievo** [comment]
Considering the issue arises in the frontend itself, I think a sensitization should suffice for number 2.

**Comment by CMCDragonkai** [comment]
> Console output from handler for number 2:
> 
> ```ts
> {
>   nodeId: IdInternal(32) [Uint8Array] [
>     211, 206, 216, 221, 137,  66, 215, 114,
>       8, 228,  98, 250, 193, 250, 150,  56,
>     210,  34,  94, 186,  71, 180, 100, 130,
>     229, 191, 254,  19,  45,  18, 227, 171
>   ],
>   address: { host: '127.0.0.1', port: 55551, hostname: undefined },
>   usageCount: 0,
>   timeout: undefined
> }
> ```
> 
> So it is more than obvious that this JSON:
> 
> ```json
> [{"host":"127.0.0.1","hostname":"","nodeIdEncoded":"vqf7dhnc98bbn4274cbtc3ukm73924nlq8uq690n5nvv16b8iselg","port":55551,"timeout":-1,"usageCount":0,"metadata":{"timeout":15000,"authorization":"Bearer {\"payload\":\"eyJpYXQiOjE2OTgyMTY4NTguNzA1LCJpc3MiOiJ2cTQ4MG9mbGQ3dnJvbDB1cmExdmlpZjk0NzRwYjdwdHJxbTlzOGxkMjN2bjV0OTJvaWY0ZyIsInN1YiI6InZxNDgwb2ZsZDd2cm9sMHVyYTF2aWlmOTQ3NHBiN3B0cnFtOXM4bGQyM3ZuNXQ5Mm9pZjRnIn0\",\"signatures\":[{\"protected\":\"eyJhbGciOiJCTEFLRTJiIn0\",\"signature\":\"UbIRKl5a8AWuNIAK2aCRmgatv0e940ZNZJ2Zvzi_WF4\"}]}"}}]
> ```
> 
> The metadata field isn't from the backend, consequently the issue lies in the frontend.

Where did you do this?

**Comment by CMCDragonkai** [comment]
Tasks should be not just fix 1, 2, 3. Summarise the titles.

File: src/nodes/CommandConnections.ts [hunk 1]
```diff
@@ -71,10 +113,18 @@ class CommandAdd extends CommandPolykey {
...
           );
         } else {
+          // Deep copy the connections array to avoid modifying the original object
+          const sanitizedConnections = JSON.parse(JSON.stringify(connections));
+          // Loop through each connection and remove the sensitive 'authorization' field
+          for (const connection of sanitizedConnections) {
+            if (connection.metadata && connection.metadata.authorization) {
+              delete connection.metadata;
+            }
+          }
```
Comment (tegefaulkes) [inline src/nodes/CommandConnections.ts:123]: This is a bandaid, not a fix. Find out why metadata is included in the first place. That's the real problem.

**Comment by addievo** [comment]
> > Console output from handler for number 2:
> > ```ts
> > {
> >   nodeId: IdInternal(32) [Uint8Array] [
> >     211, 206, 216, 221, 137,  66, 215, 114,
> >       8, 228,  98, 250, 193, 250, 150,  56,
> >     210,  34,  94, 186,  71, 180, 100, 130,
> >     229, 191, 254,  19,  45,  18, 227, 171
> >   ],
> >   address: { host: '127.0.0.1', port: 55551, hostname: undefined },
> >   usageCount: 0,
> >   timeout: undefined
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So it is more than obvious that this JSON:
> > ```json
> > [{"host":"127.0.0.1","hostname":"","nodeIdEncoded":"vqf7dhnc98bbn4274cbtc3ukm73924nlq8uq690n5nvv16b8iselg","port":55551,"timeout":-1,"usageCount":0,"metadata":{"timeout":15000,"authorization":"Bearer {\"payload\":\"eyJpYXQiOjE2OTgyMTY4NTguNzA1LCJpc3MiOiJ2cTQ4MG9mbGQ3dnJvbDB1cmExdmlpZjk0NzRwYjdwdHJxbTlzOGxkMjN2bjV0OTJvaWY0ZyIsInN1YiI6InZxNDgwb2ZsZDd2cm9sMHVyYTF2aWlmOTQ3NHBiN3B0cnFtOXM4bGQyM3ZuNXQ5Mm9pZjRnIn0\",\"signatures\":[{\"protected\":\"eyJhbGciOiJCTEFLRTJiIn0\",\"signature\":\"UbIRKl5a8AWuNIAK2aCRmgatv0e940ZNZJ2Zvzi_WF4\"}]}"}}]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The metadata field isn't from the backend, consequently the issue lies in the frontend.
> 
> Where did you do this?

@CMCDragonkai, console.log'ed the handler and ran PK with --verbose to see the output

File: src/nodes/CommandConnections.ts (hunk 1 above)
Comment (addievo) [reply src/nodes/CommandConnections.ts:123]: > This is a bandaid, not a fix. Find out why metadata is included in the first place. That's the real problem.

@tegefaulkes I think the core issue is again of rpc, which is this

    return {
      readable: outputMessageTransformStream.readable,
      writable: inputMessageTransformStream.writable,
      cancel: rpcStream.cancel,
      meta: metadata,
    };
    
The client returns the metadata input. 

How do I go about this?

Comment (tegefaulkes) [reply src/nodes/CommandConnections.ts:123]: Let's reason through this. We're getting metadata in the output and we don't know why. Where is it coming from? It is a server stream call, so we're getting a stream of the output. 

OK, so in the RPC the first message of the response, or any message can include metadata. Ok so we *do* need to strip out the metadata when reading the output.

But how do we do it? doing a `JSON.parse(JSON.stringify())` is a code smell and should almost never be done since it's a very expensive operation done twice. There are better ways to deep copy.

Finally, why not just strip out the metadata when we're collecting the stream on line 50? You don't need a conditional, you just construct a new object with the information you want.

Comment (addievo) [reply src/nodes/CommandConnections.ts]: Resolved

**Comment by CMCDragonkai** [comment]
What's the status on this PR?

**Comment by addievo** [comment]
> What's the status on this PR?

Now it is ready to merge

**Comment by tegefaulkes** [comment]
Is `nodesConnection is is sometimes lising the IP as as a ipv6 mapped ipv4 address. This needs to be looked into.` addressed from the issue?

**Comment by addievo** [comment]
> Is `nodesConnection is is sometimes lising the IP as as a ipv6 mapped ipv4 address. This needs to be looked into.` addressed from the issue?

No, this issue does not address that.

**Comment by CMCDragonkai** [comment]
That should be looked into the ipv4-mapped-ipv6 addresses.

File: src/utils/utils.ts [hunk 2]
```diff
@@ -95,20 +95,30 @@ function outputFormatter(msg: OutputObject): string | Uint8Array {
...
+
+    // Step 2: Use the maxKeyLength to align the keys and values
     for (const key in msg.data) {
       let value = msg.data[key];
       // Empty string for null or undefined values
       if (value == null) {
         value = '';
       }
-      value = JSON.stringify(value);
+      value = JSON.stringify(value).replace(/"/g, '');
```
Comment (CMCDragonkai) [inline src/utils/utils.ts:113]: This isn't safe. There may be actual `"` inside the values of the JSON. It's better to left and right trim the `"` if the value is of type string.

File: src/utils/utils.ts [hunk 3]
```diff
@@ -95,20 +95,30 @@ function outputFormatter(msg: OutputObject): string | Uint8Array {
...
       // Remove the last line terminator if it exists
-      // This may exist if the value is multi-line string
       value = value.replace(/(?:\r\n|\n)$/, '');
-      // If the string has line terminators internally
-      // Then we must append `\t` separator after each line terminator
-      value = value.replace(/(\r\n|\n)/g, '$1\t');
-      output += `${key}\t${value}\n`;
```
Comment (CMCDragonkai) [inline src/utils/utils.ts:111]: Why is the `value.replace` being removed?

**Comment by CMCDragonkai** [comment]
Can you rewrite the spec, I think some of the spec here is just plain incorrect.

**Comment by CMCDragonkai** [comment]
Not ready to be merged. Spec does not match implementation details.

File: src/nodes/CommandConnections.ts [hunk 4]
```diff
@@ -48,20 +48,69 @@ class CommandAdd extends CommandPolykey {
...
+
+            const outputLine =
+              hostString.padEnd(maxHostStringLength) +
+              '\t' +
+              usageCount.padEnd(maxUsageCountLength) +
+              '\t' +
+              timeout.padEnd(maxTimeoutLength);
             output.push(outputLine);
           }
           process.stdout.write(
```
Comment (CMCDragonkai) [inline src/nodes/CommandConnections.ts:116]: I believe `node connections` should be using a table format. Not a list format.

So this is not correct. A proper refactoring is to go to the output formatter and update the table formatting system first, and then send an array of dictionaries to the output formatter.

Focus on making the output formatter more generically useful will therefore benefit ALL commands, not just this command.

File: src/agent/CommandStatus.ts [hunk 5]
```diff
@@ -79,8 +79,6 @@ class CommandStatus extends CommandPolykey {
               clientPort: response.clientPort,
               agentHost: response.agentHost,
               agentPort: response.agentPort,
-              publicKeyJWK: response.publicKeyJwk,
-              certChainPEM: response.certChainPEM,
```
Comment (CMCDragonkai) [inline src/agent/CommandStatus.ts:83]: Didn't you say you were working on something to expand the `agent status` command to add additional details like number of connections, number of nodes in the graph, number of gestalt vertices, and sigchain claims? Where is that PR?

**Comment by addievo** [comment]
As 
https://github.com/MatrixAI/Polykey-CLI/issues/44

has been bifurcated into a new issue in Polykey, this PR is ready to be reviewed and merged.

I am doing that now.

File: src/agent/CommandStatus.ts (hunk 5 above)
Comment (addievo) [reply src/agent/CommandStatus.ts:83]: That is in a new branch called feature-cli-polish-2, since the expansion is still in discussion, I didn't want to muddy this PR.

I will be creating a PR for that soon.

File: src/nodes/CommandConnections.ts (hunk 4 above)
Comment (addievo) [reply src/nodes/CommandConnections.ts:116]: Alright, I'll get on that then.

File: src/utils/utils.ts (hunk 2 above)
Comment (addievo) [reply src/utils/utils.ts:113]: I modified the code to only remove the starting and ending quotes if the value is of type string.

File: src/utils/utils.ts (hunk 3 above)
Comment (addievo) [reply src/utils/utils.ts:111]: Added back the logic to handle /t terminators.

**Comment by CMCDragonkai** [comment]
Go to the output formatter issue and solve things there first then review this in accordance.

**Comment by addievo** [comment]
My proposed changes to the `table` type output formatter:

1. Optional header row

This would be an optional parameter which would allow adding a header row to signify which key a column represents in a particular table.

2. Adding padding based on length of longest key in a column.

3. Adding a row count indicator.

**Comment by addievo** [comment]
> Go to the output formatter issue and solve things there first then review this in accordance.

Oh okay

**Comment by addievo** [comment]
@CMCDragonkai,

So this is what I got from https://github.com/MatrixAI/Polykey-CLI/issues/22

1. TSV standardization : Output should be tab seperated values.
2. Handling special characters: Tabs and newlines should be wrapped in quotes.
3. Padding: Make a single pass thru data, and for streams, use a max len and then alter according to data being streamed.

**Comment by CMCDragonkai** [comment]
> @CMCDragonkai the issue with stream based padding, i.e. updating counter for every item in stream is that, that would have to be implemented on a case-to-case basis, since the data is fed into output formatter after it has been all *collected* and added to an array.
> 

No it can be done in the output formatter. Do a scan.

File: src/utils/utils.ts [hunk 6]
```diff
@@ -62,7 +64,93 @@ function standardErrorReplacer(key: string, value: any) {
...
+  // Initialize maxColumnLengths with header lengths if headers are provided
+  if (options?.headers) {
+    for (const header of options.headers) {
+      maxColumnLengths[header] = header.length;
+    }
+  }
+
+  let iterableStream = Array.isArray(rowStream)
+    ? arrayToAsyncIterable(rowStream)
+    : rowStream;
```
Comment (tegefaulkes) [inline src/utils/utils.ts:91]: Why is this needed? It's perfectly valid to do 

```ts
const array = [1, 2, 3];

for await (const value of array) {
  //do things
}

```

Remember, you can `await` anything, not just promises.

You don't need to convert an array to an async iterator. Unless you want to use iterator calls like `thing.next()`?

File: src/utils/utils.ts [hunk 7]
```diff
@@ -62,7 +64,93 @@ function standardErrorReplacer(key: string, value: any) {
...
+
+  const updateMaxColumnLengths = (row: TableRow) => {
+    for (const [key, value] of Object.entries(row)) {
+      const cellValue =
+        value === null || value === '' || value === undefined ? 'N/A' : value;
```
Comment (tegefaulkes) [inline src/utils/utils.ts:96]: You can simplify this. If a value is `null` or `undefined` you can use `??` to use a default.

`value ?? 'N/A`, so only `value == ''` needs to be checked. The whole thing can be compressed to `value == '' ? 'N/A' : value ?? 'N/A'`

File: src/utils/utils.ts [hunk 8]
```diff
@@ -62,7 +64,93 @@ function standardErrorReplacer(key: string, value: any) {
   return value;
 }
 
-function outputFormatter(msg: OutputObject): string | Uint8Array {
+async function* arrayToAsyncIterable<T>(array: T[]): AsyncIterable<T> {
+  for (const item of array) {
+    yield item;
+  }
+}
```
Comment (tegefaulkes) [inline src/utils/utils.ts:71]: This is a very weird thing to need.

File: src/utils/utils.ts [hunk 9]
```diff
@@ -62,7 +64,93 @@ function standardErrorReplacer(key: string, value: any) {
...
+    if (options?.includeRowCount) {
+      formattedRow += `${++rowCount}\t`;
+    }
+
+    const keysToUse = options?.headers ?? Object.keys(maxColumnLengths);
+
+    for (const key of keysToUse) {
+      const cellValue = Object.prototype.hasOwnProperty.call(row, key)
+        ? row[key]
+        : 'N/A';
```
Comment (tegefaulkes) [inline src/utils/utils.ts:135]: This is very weird. Are you just trying to convert undefined values to `'N/A'`? Just use `cellValue = row[key] ?? 'N/A';`

**Comment by tegefaulkes** [comment]
I see we strip out the `""` on strings when formatting the output. Is that something we really want? How would I tell  if `1235` is different from `"12345"` or if the whitespace was part of the value?

**Comment by CMCDragonkai** [comment]
> I see we strip out the `""` on strings when formatting the output. Is that something we really want? How would I tell if `1235` is different from `"12345"` or if the whitespace was part of the value?

You cannot. We are making a UI decision here. In fact we do space padding anyway in tables to make this work.

Actually now thinking about it, however for a "password      " to work. I imagine you'd want to show the spaces. That's a good question... and perhaps in that case we would want to show the whitespace different from padded.

One issue is then some passwords are `password   \n\n`. Right? So in that case, we cannot encode it with `\n` at all.

In fact in such cases, it's not really useful at all to simply print it with padded whitespace or otherwise.

Maybe then we should keep the double quote....

Yes let's keep the double quotes, they are the only way to know. But I can imagine an issue with things like `\t` or `\v` and other whitespace characters.

I think generally though if we are printing out a secret data, we would print it out literally. For example: singular outputs, list outputs. I believe singular outputs and list outputs is currently the same. I now realise they should be considered separate output formats.

The double quotations only come into play in the more complex outputs... That would be for list, dict, table.

**Comment by addievo** [comment]
> > I see we strip out the `""` on strings when formatting the output. Is that something we really want? How would I tell if `1235` is different from `"12345"` or if the whitespace was part of the value?
> 
> You cannot. We are making a UI decision here. In fact we do space padding anyway in tables to make this work.
> 
> Actually now thinking about it, however for a "password " to work. I imagine you'd want to show the spaces. That's a good question... and perhaps in that case we would want to show the whitespace different from padded.
> 
> One issue is then some passwords are `password \n\n`. Right? So in that case, we cannot encode it with `\n` at all.
> 
> In fact in such cases, it's not really useful at all to simply print it with padded whitespace or otherwise.
> 
> Maybe then we should keep the double quote....
> 
> Yes let's keep the double quotes, they are the only way to know. But I can imagine an issue with things like `\t` or `\v` and other whitespace characters.
> 
> I think generally though if we are printing out a secret data, we would print it out literally. For example: singular outputs, list outputs. I believe singular outputs and list outputs is currently the same. I now realise they should be considered separate output formats.
> 
> The double quotations only come into play in the more complex outputs... That would be for list, dict, table.

I have never seen a password with any whitespace character in between.t

**Comment by CMCDragonkai** [comment]
I have passwords with whitepsaces.

**Comment by addievo** [comment]
Apparently Polykey includes spaces typed while setting a password, which I think is a big UI/UX flaw.

https://security.stackexchange.com/questions/17192/why-disallow-special-characters-in-a-password

**Comment by addievo** [comment]
Also, why would we ever want to print out the password? Isn't that also a big security flaw? Printing our plaintext passwords

File: src/utils/utils.ts (hunk 6 above)
Comment (addievo) [reply src/utils/utils.ts:91]: Hey @tegefaulkes, so using `for await (const value of array)` is actually worse in terms of space complexity as you have to store the array instead of just reading it, both your approach and mine use the same time complexity of O(2n), which is what Roger wanted, so I don't see whats wrong with my implementation here. Please do elaborate.

File: src/utils/utils.ts (hunk 9 above)
Comment (addievo) [reply src/utils/utils.ts:135]: Ah right, I was specifically checking for no property, vs. checking for undefined and null as well. Changed it as such

File: src/utils/utils.ts (hunk 7 above)
Comment (addievo) [reply src/utils/utils.ts:96]: Alright, thanks, that looks much cleaner.

**Comment by CMCDragonkai** [comment]
It's not just passwords. There are other kinds of text too. I'm thinking secrets in the vaults. I think you need to add the quotes back. And then special encode any non printable characters except spaces.

**Comment by addievo** [comment]
#50

File: src/utils/utils.ts (hunk 6 above)
Comment (tegefaulkes) [reply src/utils/utils.ts:91]: I think you misunderstand, I'm saying you can consume an array and an `AsyncIterable` the same way. You don't need to convert an array into an `AsyncIterable` to consume it with a async for of loop.

Comment (amydevs) [reply src/utils/utils.ts:91]: > Hey @tegefaulkes, so using `for await (const value of array)` is actually worse in terms of space complexity as you have to store the array instead of just reading it, both your approach and mine use the same time complexity of O(2n), which is what Roger wanted, so I don't see whats wrong with my implementation here. Please do elaborate.

i don't think there is any performance difference between

```
const array = [1, 2, 3];

for await (const value of array) {
  //do things
}
```

and

```
const array = [1, 2, 3];

for (const value of array) {
  //do things
}
```

Also, `arrayToAsyncIterable` would create a AsyncIterable instance, where as  `for (const value of array)` wouldn't need to

File: src/nodes/CommandConnections.ts (hunk 1 above)
Comment (amydevs) [reply src/nodes/CommandConnections.ts]: how was this resolved @addievo ?

File: src/utils/utils.ts (hunk 8 above)
Comment (amydevs) [reply src/utils/utils.ts:71]: it's been removed

File: src/utils/utils.ts (hunk 7 above)
Comment (amydevs) [reply src/utils/utils.ts:96]: I changed this to
```ts
const cellValue = row[key] == null || row[key].length === 0 ? 'N/A' : row[key];
```
so it's a bit more readable
